Add abstracted access to Device Attestation Credentials

tcarmelveilleux opened the pull request:
#### Problem

- PR #8848 introduced some abstraction of access, but in
  a way that was not dependency injectable, and that was tied
  to the ConfigurationManager.
- Issue #8606 identifies a possible way to make per-product factory data
  more easy to abstract.

#### Change overview

- This PR introduces a true injectable/replaceable abstraction
  for device attestation credentials access, and takes the example
  data and tests from #8848 and makes them available as bootstrap
  version

Issue #8435
Issue #8606

#### Testing

Testing done: added unit tests for entire feature, and added
example to build.

woody-apple commented:
@pan-apple ?

msandstedt commented on src/credentials/DeviceAttestationCredsAccess.cpp:
```diff
@@ -0,0 +1,85 @@
+};
+
+// Default to avoid nullptr on getter and cleanly handle new products/clients before
+// their provide their own.
```
Nit:

```suggestion
// they provide their own.
```

tcarmelveilleux replied:
Done

mspang reviewed (approved):
Looks good. I would start a bikeshed of the name though - maybe accessor -> provider ? "Access" seems like it could just as easily apply to the client of this interface which suggests it may not be the most descriptive name for the interface itself.

Also this PR seems to switch between local variable naming styles.

andy31415 commented:
@chrisdecenzo @Damian-Nordic @hawk248 @jepenven-silabs @msandstedt ?

tcarmelveilleux commented:
> Looks good. I would start a bikeshed of the name though - maybe accessor -> provider ? "Access" seems like it could just as easily apply to the client of this interface which suggests it may not be the most descriptive name for the interface itself.
> 

Moved to `Provider` suffix, since it's better than `Accessor`.

> Also this PR seems to switch between local variable naming styles.

Fixed that part
